doc/source/rllib/env-to-module-connector.rst

```diff
@@ -0,0 +1,546 @@
+            [ 0.0292,  0.0259, -0.0322, -0.0004]]
+        ),
+        'state_in': {
+            # Note: The shape of each state tensor here is
+            # (B=2, [num LSTM-layers=1], [LSTM cell size]).
+            'h': tensor([[[0., 0., .., 0.]]]),
+            'c': tensor([[[0., 0., ... 0.]]]),
+        },
+    }
+
+.. hint::
+
+    You are free to design the internal states of your custom :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` classes
+    however you like. You only need to override the :py:meth:`` method and make sure you return a new state of any nested structure and
+    shape from your `forward_..()` methods under the fixed ``state_out`` key.
+    See `here for an example <https://github.com/ray-project/ray/blob/master/rllib/examples/rl_modules/classes/lstm_containing_rlm.py>`__
+    of an RLModule class with a custom LSTM layer in it.
+
+
+Writing custom env-to-module connectors
+---------------------------------------
+
+You can customize the default env-to-module pipeline that RLlib creates through specifying a function in your
+:py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`, which takes an optional RL environment object (`env`) and an optional `spaces`
+dictionary as input arguments and returns a single :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` piece or a list thereof.
+RLlib prepends the provided :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` instances to the
+:ref:`default env-to-module pipeline <default-env-to-module-pipeline>` in the order returned,
+unless you set `add_default_connectors_to_env_to_module_pipeline=False` in your config, in which case RLlib exclusively uses the provided
+:py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pieces without any automatically added default behavior.
+
+.. _env-to-module-connectors-structure-of-spaces-arg:
+
+Note that RLlib expects the structure of the `spaces` argument to be:
+
+.. code-block:: python
+
+    spaces = {
+        "__env__": ([env observation space], [env action space]),  # <- may be vectorized
+        "__env_single__": ([env observation space], [env action space]),  # <- never vectorized!
+        "[module ID, e.g. 'default_policy']": ([module observation space], [module action space]),
+        ...  # <- more modules in multi-agent case
+    }
+
+For example, to prepend a custom ConnectorV2 piece to the env-to-module pipeline, you can do this in your config:
+
+.. testcode::
+    :skipif: True
+
+    # Your builder function must accept an optional `gymnasium.Env` and an optional `spaces` dict
+    # as arguments.
+    config.env_runners(
+        env_to_module_connector=lambda env, spaces, device: MyEnvToModuleConnector(..),
+    )
+
+
+If you want to add multiple custom pieces to the pipeline, return them as a list:
+
+.. testcode::
+    :skipif: True
+
+    # Return a list of connector pieces to make RLlib add all of them to your
+    # env-to-module pipeline.
+    config.env_runners(
```

Comment:
This show case is so valuable as it is often unclear how multiple connectors should be preprended.